File: python/graph/BellmanFord.py
# -*- coding: utf-8 -*-
# 解法1　これでTLEならlistにする
import collections


class BellmanFord():
    def __init__(self):
        self.edges = []

    def add(self, u, v, d, directed=False):
        """
        u = from, v = to, d = cost
        directed = Trueのとき、有向グラフである。
        """
        if directed is False:
            self.edges.append([u, v, d])
            self.edges.append([v, u, d])
        else:
            self.edges.append([u, v, d])

    def BellmanFord_search(self, s):
        """
        次のBellmanFord_negative_boolまで考慮した時には、0-indexedであることに注意
        :param s: 始点
        :return: d[i] 始点sから各点iまでの最短経路
        """
        d = collections.defaultdict(lambda: float('inf'))
        d[s] = 0
        numEdges = len(self.edges)
        while True:
            update = False
            for i in range(numEdges):
                e = self.edges[i]
                # e: 辺iについて [from,to,cost]
                if d[e[0]] != float("inf") and d[e[1]] > d[e[0]] + e[2]:
                    d[e[1]] = d[e[0]] + e[2]
                    update = True
            if not update:
                break
        return d

    def BellmanFord_negative_bool(self, numNodes, zero_indexed=False):
        # 負の閉路の検出, Trueなら負の閉路が存在する
        d = collections.defaultdict(lambda: float('inf'))
        d[0] = 0
        numEdges = len(self.edges)
        for i in range(numNodes):
            for j in range(numEdges):
                e = self.edges[j]
                if d[e[1]] > d[e[0]] + e[2]:
                    d[e[1]] = d[e[0]] + e[2]
                    if i == numNodes-1:
                        return True
        return False

# 解法2


class BellmanFord():
    def __init__(self, N):
        self.N = N
        self.edges = []

    def add(self, u, v, d, directed=False):
        """
        u = from, v = to, d = cost
        directed = Trueのとき、有向グラフである。
        """
        if directed is False:
            self.edges.append([u, v, d])
            self.edges.append([v, u, d])
        else:
            self.edges.append([u, v, d])

    def BellmanFord_search(self, s):
        """
        :param s: 始点
        :return: d[i] 始点sから各点iまでの最短経路
        """
        d = [float('inf') for i in range(self.N)]
        d[s] = 0
        numEdges = len(self.edges)
        while True:
            update = False
            for i in range(numEdges):
                e = self.edges[i]
                # e: 辺iについて [from,to,cost]
                if d[e[0]] != float("inf") and d[e[1]] > d[e[0]] + e[2]:
                    d[e[1]] = d[e[0]] + e[2]
                    update = True
            if not update:
                break
        return d

    def BellmanFord_negative_bool(self, numNodes):
        # 負の閉路の検出, Trueなら負の閉路が存在する
        d = [float('inf') for i in range(self.N)]
        d[0] = 0
        numEdges = len(self.edges)
        for i in range(numNodes):
            for j in range(numEdges):
                e = self.edges[j]
                if d[e[1]] > d[e[0]] + e[2]:
                    d[e[1]] = d[e[0]] + e[2]
                    if i == numNodes-1:
                        return True
        return False

File: rust/graph/bellmanford.rs
const INF: i64 = 1 << 61;


// adjacent list
struct Graph {
    edges: Vec<(usize, usize, i64)>, // adjacent list
    n: usize
}

impl Graph {
    fn new(n: usize) -> Self {
        Graph {
            edges: vec![],
            n: n
        }
    }

    fn add_costs(&mut self, from: usize, to: usize, cost: i64, directed: bool) {
        if directed {
            self.edges.push((from, to, cost));
        } else {
            self.edges.push((from, to, cost));
            self.edges.push((to, from, cost));
        }
    }

    fn bellmanford_search(&self, start: usize) -> Vec<i64> {
        let mut dist: Vec<i64> = vec![INF; self.n];
        dist[start] = 0;
        loop {
            let mut update: bool = false;
            for edge in &self.edges {
                let (from, to, cost) = *edge;
                if (dist[from] != INF) & (dist[to] > dist[from] + cost) {
                    dist[to] = dist[from] + cost;
                    update = true;
                }
            }
            if !update {
                break;
            }
        }
        return dist
    }

    fn bellmanford_search_negative_cycle(&self, start: usize) -> (bool, Vec<i64>) {
        // (true, dist) なら負の閉路を含む
        let mut dist: Vec<i64> = vec![INF; self.n];
        dist[start] = 0;
        for i in 0..self.n {
            for &(from, to, cost) in &self.edges{
                if (dist[from] != INF) & (dist[to] > dist[from] + cost) {
                    dist[to] = dist[from] + cost;
                    if i == self.n - 1{
                        return (true, dist);
                    }
                }
            }
        }
        return (false, dist);
    }
}

//adjacent matrix
struct Graph {
    edges: Vec<Vec<(usize, i64)>>, // adjacent matrix
    n: usize
}

impl Graph {
    fn new(n: usize) -> Self {
        Graph {
            edges: vec![Vec::new(); n],
            n: n
        }
    }

    fn add_costs(&mut self, from: usize, to: usize, cost: i64, directed: bool) {
        if directed {
            self.edges[from].push((to, cost));
        } else {
            self.edges[from].push((to, cost));
            self.edges[to].push((from, cost));
        }
    }

    fn bellmanford_search(&self, start: usize) -> Vec<i64> {
        let mut dist: Vec<i64> = vec![INF; self.n];
        dist[start] = 0;
        let mut update: bool = true;
        while update {
            update = false;
            for from in 0..self.n {
                for edge in &self.edges[from] {
                    let (to, cost) = *edge;
                    if dist[to] > dist[from] + cost {
                        dist[to] = dist[from] + cost;
                        update = true;
                    }
                }
            }
        }
        return dist
    }

    fn bellmanford_search_negative_cycle(&self, start: usize) -> (bool, Vec<i64>) {
        // (true, dist) なら負の閉路を含む
        let mut dist: Vec<i64> = vec![INF; self.n];
        dist[start] = 0;
        let mut counter = 0;
        let mut update: bool = true;
        while update {
            update = false;
            for from in 0..self.n {
                for edge in &self.edges[from] {
                    let (to, cost) = *edge;
                    if (dist[from] != INF) & (dist[to] > dist[from] + cost) {
                        dist[to] = dist[from] + cost;
                        update = true;
                        counter += 1;
                    }
                }
            }
            if counter >= self.n * self.n {
                return (true, dist);
            }
        }
        return (false, dist);
    }
}